### ISSUE TITLE
Official process to read cookies on connection

### ISSUE THREAD

**Issue by DrLightman** [socketio/socket.io#2470]
Hello,

I'm new to both node and socket.io and I'm quite struggling to fetch informations from documentation.

I added socket.io to my web app where users are authenticated via a cookie token which destroys and renovates itself at every new login. This client web app (php+html+js) now connects to a server (node) and in the server I have to authenticate user's socket connections.

I gladly noticed that on connection cookies are transmitted inside:

``` javascript
io.on( 'connection', function ( socket ) {
  // -> socket.handshake.headers.cookie string
  // -> I bound to socket.id
} );
```

And I'm using node's "cookie" module to parse them than looking up the db with "mysql" module to verify it and couple with socket.id.

My doubts are:
1. does an official method/function exist to access socket.handshake.headers.cookie object or part of it and can I be sure it will be ever there? Now I'm just testing with an if if properties exists. if( 'handshake' in socket && ... ) ...
2. subsequents socket's emits don't transmit cookies, do they? eg there is no way for me to read cookies at this point:

``` javascript
    socket.on( 'disconnect', function ( data ) {
        console.log( 'someone disconnected', data );
    } );
```

Thank you

**Comment by LordMajestros** [comment]
1. The official way is socket.request.headers.cookie from 1.4.x and socket.handshake.headers.cookie in 1.3.x
2. Yes subsequent socket emits do not transmit cookies. The only way to read cookies at this point is from `socket.handshake.headers.cookie` or `socket.request.headers.cookie` or by storing them somewhere in 
```
io.on( 'connection', function ( socket ) {
  // socket.cookie = socket.handshake.headers.cookie || socket.request.headers.cookie
} );
```

**Comment by DrLightman** [comment]
Excellent, thank you.

My doubt at this point is what happens when the cookie token changes in the client dued a logout and re-login. (it's a security feature of the client to change the token at every login)

My web client (full ajax) has no page reloads that would trigger a socket disconnect and consequent on/connection on the server to re-authenticate the socket.

But now that it comes in mind, I think I can manage with the coupling between socket-id<->token in the database to verify this at every emits. Since I reneovated the toekn, I destroyed the preovious entry socket.id<->token so it doesn't verify no more and I can issue a socket disconnect an re connect server side.
